pkg/analyzer/nodes/node.go

```diff
@@ -0,0 +1,211 @@
+		if err := n.addFiles(filepath.FromSlash(dir), matcher, &m); err != nil {
+			return nil, errors.Wrapf(err, "failed to match filesMap at %q (at %q)", dir, n.root)
+		}
+	}
+	m.Emit()
+	return n, nil
+}
+
+// addFiles searches the directory for files using the matcher and adds them to the file map.
+func (n *filesMap) addFiles(dir string, matcher matcher.Matcher, m *metrics.FileExtractionMetrics) error {
+	logrus.WithFields(logrus.Fields{
+		"root":      n.root,
+		"directory": dir,
+	}).Info("add files from directory")
+	return filepath.WalkDir(filepath.Join(n.root, dir), func(fullPath string, entry fs.DirEntry, err error) error {
+		if err != nil {
+			if filesIsNotAccessible(err) {
+				m.ErrorCount()
+				return nil
+			}
+			return errors.Wrapf(err, "failed to read %q", fullPath)
+		}
+		relPath, err := filepath.Rel(n.root, fullPath)
+		if err != nil {
+			return errors.Wrapf(err, "walking %q", dir)
+		}
+		path := filepath.ToSlash(relPath)
+		f, err := extractFile(path, entry, matcher, m)
+		if err != nil {
+			logrus.WithFields(logrus.Fields{
+				"root":          n.root,
+				"directory":     dir,
+				"relative_path": relPath,
+			}).Error("extract file")
+			return err
+		}
+		if f != nil {
+			n.files[path] = f
+		}
+		return nil
+	})
+}
+
+// filesIsNotAccessible returns true if the error means the file cannot be read.
+func filesIsNotAccessible(err error) bool {
+	return errors.Is(err, fs.ErrNotExist) || errors.Is(err, fs.ErrPermission)
+}
+
+// extractFile extracts data from the given directory entry, if granted by the
+// path matcher, and if it passes safety checks. Otherwise, returns nil.
+func extractFile(path string, entry fs.DirEntry, pathMatcher matcher.Matcher, m *metrics.FileExtractionMetrics) (*fileData, error) {
+	// Ignore all directories.
+	if entry.IsDir() {
+		return nil, nil
+	}
+	fileInfo, err := entry.Info()
+	if err != nil {
+		if filesIsNotAccessible(err) {
+			m.ErrorCount()
+			return nil, nil
+		}
+		// We assume that other errors reflect underlying problems with the filesystem
+		// worth failing the whole extraction.
+		return nil, err
+	}
+	m.FileCount()
+	// FIXME No matcher actually uses the file contents, except the matcher for
+	//       language components, which is not currently used by directory scans, hence we
+	//       offer nil.  This will break if a language component matcher (analyzer) is used.
+	matches, isExtractable := pathMatcher.Match(path, fileInfo, nil)
+	if !matches {
+		return nil, nil
+	}
+	m.MatchCount()
+	// File size limit check.
+	if isExtractable && fileInfo.Size() > analyzer.GetMaxExtractableFileSize() {
+		logrus.Errorf("skipping file %q (%d bytes): size is greater than maxExtractableFileSize of %d MiB",
+			path, fileInfo.Size(), analyzer.GetMaxExtractableFileSize()/1024/1024)
+		return nil, nil
+	}
+	return &fileData{
+		isExecutable:  matcher.IsFileExecutable(fileInfo),
+		isExtractable: isExtractable,
+	}, nil
+}
+
+// Get implements analyzer.Files
```

Comment:
Can you explain more the function of Get function?

Reply:
The details are explained in the documentation of `analyzer.Files`. If you have a better way of linking them together let me know.